### PR TITLE
Adding TinyMCE Wysiwyg Field Type

### LINKED ISSUE #928: Add TinyMCE Wysiwyg Field
Separate to #616, I want to add a WYSIWYG html field for the Admin UI.

It'll be useful while the content field is under development, and longer term helpful for transitioning from Keystone 4 or any other CMS that stores content in HTML.

The initial version should be a basic implementation of [TinyMCE 5](https://www.tiny.cloud/) and we should skin it a bit to fit in with the Admin UI style like we did in Keystone 4.

### PR DESCRIPTION
**Pull request by JedWatson**
### Ready for Review

Closes #928 

This adds a new field type package that renders a TinyMCE Editor in the Admin UI. Behind the scenes it inherits the implementation from the `Text` field.

There are a few cleanup tasks, which I'll open a separate issue for.

---

### Original Content

@mitchellhamilton I'm having trouble getting the `preconstruct` setup here to work. Would you mind taking a look to see if you can work out what I've missed, or if there's an underlying issue?

I get this error:

```
$ preconstruct build
🎁 info building bundles!
🎁 error @keystone-alpha/fields-wysiwyg-tinymce/views/Field main field is invalid
```

However I **think** I've set it up the same way you have for the entry points in the `fields` package. The paths look valid:

```json
{
  "main": "dist/fields.cjs.js",
  "module": "dist/fields.esm.js",
  "preconstruct": {
    "source": "./Field"
  }
}
```

### CODE REVIEW

**Comment by mitchellhamilton** [comment]
The dist file names are derived from the package name(it's the package name but if it has a scope, the scope is removed) so for this, the filenames would be `dist/fields-wysiwyg-tinymce.cjs.js` and `dist/fields-wysiwyg-tinymce.esm.js`

This problem can also be fixed by running `yarn preconstruct fix` but I just checked to make sure it would work and noticed that it added a newline to every entrypoint package.json(even though it didn't change them in any other way) which isn't a great experience so I fixed that and did a release so if you update to `preconstruct@0.0.51` and then run `yarn preconstruct fix`, it'll fix it for you.

There are two things that I want to do in preconstruct to make this a better experience in the future for other people.

1. Show the incorrect value for the field and what the value should be
2. For any errors that are fixable with `preconstruct fix`, say that the error is fixable with preconstruct fix below the error message

File: packages/fields-wysiwyg-tinymce/package.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "@keystone-alpha/fields-wysiwyg-tinymce",
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/package.json]: Bikeshed: should the name be singular, since there is only one field in this package?

```
"name": "@keystone-alpha/field-wysiwyg-tinymce",
```

File: packages/fields-wysiwyg-tinymce/views/Field/Field.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            justifyContent: 'space-between',
+          }}
+        >
+          {field.label}{' '}
+          {!canRead ? (
+            <ShieldIcon title={error.message} css={{ color: colors.N20, marginRight: '1em' }} />
+          ) : null}
+        </FieldLabel>
+        <FieldInput>
+          <Input
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/views/Field/Field.js]: This is where you'll put tinyMCE, right?

**Comment by gautamsi** [comment]
https://github.com/keystonejs/keystone-5/issues/928#issuecomment-476346462
> How about a provider/adapter system with this field type? I am sure people would want to use something like Trumbowyg and Quill     
if there is adapter/provider system, dev can add their own editor without creating whole new FieldType

File: packages/fields-wysiwyg-tinymce/package.json (hunk 1 above)
Comment (JedWatson) [reply packages/fields-wysiwyg-tinymce/package.json]: Good question. I went back and forth myself on the package name a few times.

Settled on `fields` partly just becuase it makes the structure read more coherently in the `packages` directory. Also because some packages may contain multiple field types (e.g I expect `CloudinaryImage` and `CloudinaryImages` would go in the same `fields-cloudinary` package) so it might be nice to keep the naming convention consistent.

**Comment by JedWatson** [comment]
> The dist file names are derived from the package name(it's the package name but if it has a scope, the scope is removed) so for this, the filenames would be dist/fields-wysiwyg-tinymce.cjs.js and dist/fields-wysiwyg-tinymce.esm.js

Right! ok that makes sense. If you can detect this is why it failed, might be nice to explain that in the error itself (or link to docs, etc)... was a bit of a dead end otherwise. Maybe even hint to run `yarn preconstruct fix`?

Related: I actually nearly went through renaming the dist file names for the entries in the `fields` package so they matched the field type and entry names (e.g `Text.Cell.esm.js` would make more sense than `fields.esm.js` for the `fields/types/Text/views/Cell` entry. 🤷‍♂️ maybe we should relax this restriction or something?

Anyway, thanks! looks like it worked. Also, this was really nice:

```
🎁 ?    @keystone-alpha/fields-wysiwyg-tinymce Babel helpers (functions inserted by babel transforms) should be imported from a @babel/runtime {...}
```

**Comment by mitchellhamilton** [comment]
>Right! ok that makes sense. If you can detect this is why it failed, might be nice to explain that in the error itself (or link to docs, etc)... was a bit of a dead end otherwise. Maybe even hint to run yarn preconstruct fix?

Yup, that's the plan.

> Related: I actually nearly went through renaming the dist file names for the entries in the fields package so they matched the field type and entry names (e.g Text.Cell.esm.js would make more sense than fields.esm.js for the fields/types/Text/views/Cell entry. 🤷‍♂️ maybe we should relax this restriction or something?

I definitely think there needs to be some way to have different filenames, especially between entrypoints in a package. I don't think it should just be letting users customize it directly though since enforcing it enables really awesome things like auto fixing and initializing, I'm not entirely sure what it should look like yet though. Right now, what I'm thinking of is the name that it is currently + path to entrypoint but `/` replaced with `-`, so for example for `fields/types/Text/views/Cell` making it, `fields-types-Text-views-Cell.esm.js`. The other thing I'm thinking of is letting people specify a name in entrypoints just like you can in packages and using that name for the file name rather than the package's name so the name could be `text.cell`(lowercase to follow npm package naming requirements to make things consistent with when you have a single entrypoint which is the root of the package and so the name comes from the package's name) and a filename would be `text.cell.esm.js`

**Comment by JedWatson** [comment]
@gautamsi 

> How about a provider/adapter system with this field type? I am sure people would want to use something like Trumbowyg and Quill. If there is adapter/provider system, dev can add their own editor without creating whole new FieldType

At this point in the PR, I have TinyMCE working as a field in the Admin UI. If you take a look at the code, you'll see that there's _very little to this package that isn't specific to implementing TinyMCE_.

That's how we've designed field types; easy to change just the minimum when creating you own, and easy to inherit from other types. In this case I'm using most of the functionality of the `Text` field and just customising its view to render a TinyMCE editor.

So I think that's what you're actually asking for here, and adding an adapter/provider pattern would just introduce an unnecessary additional layer of complexity.

Do let me know if you think it could be simplified further though!

**Comment by JedWatson** [comment]
So this is ready for review!

It works, and presents pretty good basic functionality for WYSIWYG HTML in Keystone.

image

There are a few cleanup tasks I'd like to do (like making the toolbar and plugins customisable through options) that I'll create a followup issue for.

File: packages/fields-wysiwyg-tinymce/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+In addition to using this field type in a List, for it to work in the Admin UI you'll need to serve the TinyMCE assets from your webserver.
+
+The Field includes a function that sets up the middleware for you. Configure it in your custom server like this:
+
+```js
+const keystone = require('@keystone-alpha/core');
+const WysiwygField = require('@keystone-alpha/fields-wysiwyg-tinymce');
+
+keystone.prepare().then(async ({ server }) => {
+  WysiwygField.bindStaticMiddleware(server);
```
Comment (mitchellhamilton) [inline packages/fields-wysiwyg-tinymce/README.md]: Why is this necessary? It feels really annoying to have to do this because we're breaking out of the API of field types which people are used to.

File: packages/fields-wysiwyg-tinymce/package.json (hunk 1 above)
Comment (jesstelford) [reply packages/fields-wysiwyg-tinymce/package.json]: Even in the `Cloudinary` example, it still feels singular to me; "This field is related to `Cloudinary`.

BUT! I don't really care all that much. We just need to make sure we're consistent going forward 👍

File: packages/fields-wysiwyg-tinymce/README.md (hunk 3 above)
Comment (jesstelford) [reply packages/fields-wysiwyg-tinymce/README.md]: I'm ok with this. It's not quite a _field_, but it's a part that's necessary for the field to work. Really, fields can export whatever they like, as long as their main export includes the minimum stuff for KS5 to understand it. If it exports more; so be it 👍

File: packages/fields-wysiwyg-tinymce/README.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
+---
+section: field-types
+title: Wysiwyg
+---
+
+# WYSIWYG HTML Field powered by TinyMCE Editor
+
+This field inserts a string path into your schema based on the `Text` field type implementation, and renders a WYSIWYG editor in the Admin UI using [TinyMCE](https://www.tiny.cloud/)
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/README.md]:
```suggestion
This field inserts a string path into your schema, and renders a WYSIWYG editor in the Admin UI using [TinyMCE](https://www.tiny.cloud/)
```

The internal implementation probably isn't really important to most users.

File: packages/fields-wysiwyg-tinymce/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+keystone.prepare().then(async ({ server }) => {
+  WysiwygField.bindStaticMiddleware(server);
+  await server.start();
+});
+```
+
+```DOCS_TODO
+TODO
+```
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/README.md]: Can remove this `TODO` (the docs look pretty complete to me based on the current implementation 👍 )

File: packages/fields-wysiwyg-tinymce/index.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const { Text } = require('@keystone-alpha/fields');
+const path = require('path');
+
+function bindStaticMiddleware(server) {
+  const tinymce = require.resolve('tinymce');
+  const tinymcePath = tinymce.substr(0, tinymce.lastIndexOf('/'));
+  server.app.use('/tinymce', server.express.static(tinymcePath));
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/index.js]: This doesn't take into account any `adminPath` the user may have set (for example; in the blog project, it's set to `/admin` (the default)).

This is non-blocking. Have added it to the list of follow-up tasks in #936.

File: packages/fields-wysiwyg-tinymce/views/Editor.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+// Tell TinyMCE where its assets are
+tinymce.baseURL = '/tinymce';
+
+const defaultOptions = {
+  // min_height: 80,
+  autoresize_bottom_margin: 20,
+  branding: false,
+  menubar: false,
+  // toolbar_drawer: 'floating',
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/views/Editor.js]:
```suggestion
```

File: packages/fields-wysiwyg-tinymce/views/Editor.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
+import React, { useEffect, useRef } from 'react';
+import { Global, css } from '@emotion/core';
+import PropTypes from 'prop-types';
+import tinymce from 'tinymce/tinymce';
+
+// Tell TinyMCE where its assets are
+tinymce.baseURL = '/tinymce';
+
+const defaultOptions = {
+  // min_height: 80,
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/views/Editor.js]:
```suggestion
```

File: packages/fields-wysiwyg-tinymce/views/Editor.js [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+          editor.on('change keyup setcontent', handleChange);
+        });
+      },
+    };
+    tinymce.init(options);
+
+    return () => {
+      tinymce.remove(editorRef.current);
+    };
+  }, []);
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/views/Editor.js]:
```suggestion
  }, [this.props.autoFocus, this.props.plugins, this.props.toolbar, this.props.value, this.props.isDisabled]);
```

These are external variables which the editor relies on - we'd want to remount it if they change.

File: packages/fields-wysiwyg-tinymce/views/Field/Field.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            display: 'flex',
+            flexDirection: 'row',
+            justifyContent: 'space-between',
+          }}
+        >
+          {field.label}
+        </FieldLabel>
+        <div css={{ display: 'flex', flex: 1 }}>
+          <Editor
+            value={canRead ? value : undefined}
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/views/Field/Field.js]:
```suggestion
            value={value}
```

Line 25 means this can never be false at this point.

File: packages/fields-wysiwyg-tinymce/views/Editor.js [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  useEffect(() => {
+    if (!editorIsReady() || typeof value !== 'string' || value === contentRef.current) return;
+    editorRef.current.setContent(value);
+  }, [value]);
+
+  // Update the Editor mode when the isDisabled prop changes
+  useEffect(() => {
+    if (!editorIsReady()) return;
+    editorRef.current.setMode(isDisabled ? 'readonly' : 'design');
+  }, [isDisabled]);
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/views/Editor.js]:
```suggestion
  }, [isDisabled, editorRef.current]);
```

If the editor remounts, we want to re-apply the `isDisabled` option.

File: packages/fields-wysiwyg-tinymce/views/Editor.js [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    return () => {
+      tinymce.remove(editorRef.current);
+    };
+  }, []);
+
+  // Update the Editor content when the value prop changes
+  useEffect(() => {
+    if (!editorIsReady() || typeof value !== 'string' || value === contentRef.current) return;
+    editorRef.current.setContent(value);
+  }, [value]);
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/views/Editor.js]:
```suggestion
  }, [value, contentRef.current, editorRef.current]);
```

If the editor was remounted, the ref would change, but the value would not get updated correctly (I think). This change should fix that.

File: packages/fields-wysiwyg-tinymce/views/Field/Field.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
+/* eslint-disable import/no-extraneous-dependencies */
+/** @jsx jsx */
+
+import { jsx } from '@emotion/core';
+import { Component } from 'react';
+
+import { FieldContainer, FieldLabel } from '@arch-ui/fields';
+import { ShieldIcon } from '@arch-ui/icons';
+import { colors } from '@arch-ui/theme';
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/views/Field/Field.js]:
```suggestion
```

This isn't used below.

File: packages/fields-wysiwyg-tinymce/views/Field/Field.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
+/* eslint-disable import/no-extraneous-dependencies */
+/** @jsx jsx */
+
+import { jsx } from '@emotion/core';
+import { Component } from 'react';
+
+import { FieldContainer, FieldLabel } from '@arch-ui/fields';
+import { ShieldIcon } from '@arch-ui/icons';
```
Comment (jesstelford) [inline packages/fields-wysiwyg-tinymce/views/Field/Field.js]:
```suggestion
```

This isn't used below.

File: packages/fields-wysiwyg-tinymce/README.md (hunk 3 above)
Comment (JedWatson) [reply packages/fields-wysiwyg-tinymce/README.md]: So this was really tricky because TinyMCE basically *needs* to be able to load files from its package at runtime. We either have to get users to sign up for a `tiny.cloud` account and provide an API key, or we serve them ourselves from _somewhere_.

Basically you just can't vendor this stuff from an npm module using webpack, there are just too many issues that are fundamental to TinyMCE's architecture and history that prevent that from being possible - at least as far as they've got in v5. I spent _hours_ trying not to need this and tl;dr failed.

(to be clear; their official recommendation in the docs is to _literally copy the files when you download the package, into somewhere in your repo that static assets are hosted_. This is not something I want to do, so consider the solution here a middleground.)

Now that the field is decoupled from keystone, the admin ui, and the webserver, it's harder than it was in Keystone 4's monolith (in a good way) to serve random static resources. I was actually really happy with this solution, because it's a neat solution to an otherwise horrid problem.

If the field package exports its own bindings, and they're easy to plug into the server, it means everything is still encapsulated and we don't have issues with updating tinymce / etc like we have in the past.

File: packages/fields-wysiwyg-tinymce/index.js (hunk 6 above)
Comment (gautamsi) [reply packages/fields-wysiwyg-tinymce/index.js]: Should we make it easy for migration, use `static` path from config? (similar to v4)

File: packages/fields-wysiwyg-tinymce/README.md (hunk 3 above)
Comment (mitchellhamilton) [reply packages/fields-wysiwyg-tinymce/README.md]: Could you add a comment basically saying what you said here for people looking at this in the future?

File: packages/fields-wysiwyg-tinymce/README.md (hunk 4 above)
Comment (JedWatson) [reply packages/fields-wysiwyg-tinymce/README.md]: I actually think this is useful information - and I'd like to link to the `Text` documentation because that will eventually describe the graphql schema / filters / etc that are relevant for this field as well.

File: packages/fields-wysiwyg-tinymce/README.md (hunk 5 above)
Comment (JedWatson) [reply packages/fields-wysiwyg-tinymce/README.md]: I want to add a few more things, so leaving this in place for now (task in #936)

File: packages/fields-wysiwyg-tinymce/views/Editor.js (hunk 9 above)
Comment (JedWatson) [reply packages/fields-wysiwyg-tinymce/views/Editor.js]: I'm OK with this for a few reasons:

* `autoFocus` only matters when the component first mounts
* `value` and `isDisabled` have separate effect hooks that handle when they change
* have added `plugins` and `toolbar` for completeness, but fwiw those are configured in the options of the field when the list is created. so they can never change without a server restart.

File: packages/fields-wysiwyg-tinymce/views/Editor.js (hunk 11 above)
Comment (JedWatson) [reply packages/fields-wysiwyg-tinymce/views/Editor.js]: 👌 nice catch!

File: packages/fields-wysiwyg-tinymce/views/Editor.js (hunk 12 above)
Comment (JedWatson) [reply packages/fields-wysiwyg-tinymce/views/Editor.js]: If the editor re-mounts, the content is set when it initialises. And the synchronisation between content and props is pretty delicate, this introduces an unnecessary update. Going to leave as-is

**Comment by JedWatson** [comment]
Thanks for the comprehensive review @jesstelford!

I think this is good to go now?

File: packages/fields-wysiwyg-tinymce/README.md (hunk 3 above)
Comment (gautamsi) [reply packages/fields-wysiwyg-tinymce/README.md]: My concept in #842  would help here.  basically postponing field initialization and server preparation to the very end of processing queue. This way we can have `keystone`/`server` instance passed to Field type at the initialization, which would remove all this complexity.

We have to pass keystone/server to `list.initFields()`

File: packages/fields-wysiwyg-tinymce/README.md (hunk 4 above)
Comment (jesstelford) [reply packages/fields-wysiwyg-tinymce/README.md]: > that will eventually describe the graphql schema / filters / etc that are relevant for this field as well.

Great point 👍

**Comment by JedWatson** [comment]
I _believe_ that the tests here are failing because of an unrelated reason. It's a cypress test that's also failing on master, so I'm going to merge this in 👍

**Comment by JedWatson** [comment]
Scratch that, I ran them again and they went green!

We have test gremlins 😬
